changefeedccl: Update sink metrics

miretskiy opened the pull request:
See commits for details.

Move sink related metrics down to sink level.
Replace meaningless "processing time" metric w/ the metric keep track of the amount
of time event spent queued.
Add `sli_scope` support to changefeeds;

CRDB-2422

Release Notes: Update and improve the set of sink specific changefeed metrics
Release Notes (enterprise change): Changefeeds can be created with a new
option called `sli_scope` which lets operators configure changefeeds to use
dedicated set of metrics for those changefeed(s) so that they can be monitored
independently of other changefeed(s) in the system.

Commit:
changefeedccl: Replace processing nanos metric with a queue time metric.

Processing time metric has been broken for quite some time.  It's almost
always 0 -- because we set the "get" time to the current time, and immediately
after we increment "processing time" counter.

Instead, track the amount of time the event spent queued in the blocking buffer.

Release Notes: None

Commit:
changefeedccl: Push sink related metrics down to sink level.

Remove metric sink wrapper, and instead, push sink related metrics
directly to sink implementations.  The sinks themselves are responsible
for updating metrics.

This is done because sinks have different batching characteristics.
Incrementing counters s.a. such as latency histograms, might not be accurate
since the message is not emitted until the batch has been flushed.

This change also modifies and updates the set of metrics we keep track of.
  * Emitted message metrics (bytes and counts) are now updated when those
    messages are flushed (as opposed to buffered).
  * Metrics keeping track of emit times have been removed.  Those
    were never useful since they keep track of the
    amount of time it takes to buffer the message, which is almost always 0.
    Sometimes a message emit event may cause write to external storage (and that
    may show up in the metrics), but this event is rare, relative to the frequency
    of emit events, and thus would tend to be downsampled.
  * New histogram `changefeed.batch_hist_nanos` introduced (instead of emit histogram)
    to keep track of the amount of time the batch stays buffered in the sink before
    being flushed.  Some sinks (kafka at this time) keep track of per-message time
    instead of per-batch time.

Release Notes: Update and improve the set of sink specific changefeed metrics.

Commit:
changefeedccl: Add MVCC timestamp argument to sink EmitRow method.

Add a MVCC timestamp argument to the EmitRow call.
This change will enable tracking of various latency related metrics.

Release Notes: None

Commit:
changefeedccl: Add admit and commit latency metrics.

Add admit and commit latency histograms to changefeed metrics.
Admit latency keeps track of the differentce between MVCC timestamp
when record was updated and the time it was admitted for changefeed
processing.

The commit latency histogram keeps track of the difference between
MVCC timestamp and the time the event was acknowledged by downstream sink.

Release Notes: Add admit and commit latency metrics to changefeeds.

miretskiy commented:
Informs #70920

Commit:
changefeedccl: Add ability to track per-changefeed metrics.

Changefeeds can now be created with a new option: `metrics_label`.

Metrics label allow the operator to group metrics related to
service level indicators (SLI) by attaching a label to those metrics.
Examples of SLI related metrics are:
  * Error rates
  * Number of emitted messages and the size of data (pre & post compression).
  * Admit and commit message latency

This experimental feature is controlled via an environment variable
`COCKROACH_EXPERIMENTAL_ENABLE_PER_CHANGEFEED_METRICS` which can be set to
true to enable this feature.

While the admistrator is free to create as many of those scoped labels
as they need (up to a hard limit of 1024), excessive number of labels should
be avoided since each custom label increases the cardinality of the metrics
exported to prometheus.

SLI scopes let operators label one or more changefeeds, based on criteria
of their choosing.  This labeling effectively creates a monitoring scope,
based on relative priority of those changefeeds, which lets the operator
monitor those changefeed independently of other changefeeds
For example, production changefeed might fall under `prod` scope,
while perhaps less important, changefeeds might fall under `low_pri` scope.

The use of `metrics_label` is supported by enterprise changefeeds only.

Release Notes (enterprise change): Changefeeds can be created with a new
option called `metrics_label` which lets operators configure changefeeds to use
dedicated set of metrics for those changefeed(s) so that they can be monitored
independently of other changefeed(s) in the system.

HonoreDB reviewed (approved):
> Reviewed 1 of 20 files at r1, 1 of 13 files at r2, 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[-- commits, line 79 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Mna4GJk337IzkA7DitA):*
Typo in release notes: independenly -> independently
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 507 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Mna8CIh81EAr8P8dj0v:-Mna8CIh81EAr8P8dj0w:b-ti51w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab2451c149ef9cd51f05995490d5837f8f01d7f3/pkg/ccl/changefeedccl/changefeed_processors.go#L507)):*
> ```Go
> 	if event.Type() == kvevent.TypeKV && event.BackfillTimestamp().IsEmpty() {
> 		ca.metrics.AdmitLatency.RecordValue(timeutil.Since(event.Timestamp().GoTime()).Nanoseconds())
> 		if ca.sliMetrics != nil {
> ```

Non-blocking, but it might be nice to DRY this pattern of "call this method on global metrics, then call the same method with the same arguments on SLI metrics if they exist." Say if we had a slice of all applicable metrics to iterate over, or something like that.
___
*[pkg/ccl/changefeedccl/metrics.go, line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Mna9jCE1m9G78a5S5r0:-Mna9jCE1m9G78a5S5r1:b-41ld9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab2451c149ef9cd51f05995490d5837f8f01d7f3/pkg/ccl/changefeedccl/metrics.go#L254)):*
> ```Go
> 	metaChangefeedBatchHistNanos := metric.Metadata{
> 		Name:        "sink_batch_hist_nanos",
> 		Help:        "Time spent batched in the sink buffer before being being flushed and acknowledged",
> ```

typo: "being being"
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 334 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Mna6kNo6tQgpgkFsuj5:-Mna6kNo6tQgpgkFsuj6:b26rzin) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab2451c149ef9cd51f05995490d5837f8f01d7f3/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L334)):*
> ```Go
> 	if f.onBackfillCallback != nil {
> 		f.onBackfillCallback(scanTime)
> 		defer f.onBackfillCallback(hlc.Timestamp{})
> ```

To ensure we decrement the backfilling metric no matter how we exit. Cool.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @stevendanna)

___
*[-- commits, line 79 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Mna4GJk337IzkA7DitA):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Typo in release notes: independenly -> independently
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 507 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Mna8CIh81EAr8P8dj0v:-MnaS-_S6owLO_EgetKh:b-ubxgt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab2451c149ef9cd51f05995490d5837f8f01d7f3/pkg/ccl/changefeedccl/changefeed_processors.go#L507)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Non-blocking, but it might be nice to DRY this pattern of "call this method on global metrics, then call the same method with the same arguments on SLI metrics if they exist." Say if we had a slice of all applicable metrics to iterate over, or something like that.
</blockquote></details>

Yeah... I decided to go the "long" way for this for now... While we learn how to really use those.
Also, majority is limited to metric sink -- so the blast radius is not that large.
___
*[pkg/ccl/changefeedccl/metrics.go, line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Mna9jCE1m9G78a5S5r0:-MnaT4H30N0J_u7KT-A3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab2451c149ef9cd51f05995490d5837f8f01d7f3/pkg/ccl/changefeedccl/metrics.go#L254)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

typo: "being being"
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 334 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Mna6kNo6tQgpgkFsuj5:-MnaT5b45no6SFipORis:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab2451c149ef9cd51f05995490d5837f8f01d7f3/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L334)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

To ensure we decrement the backfilling metric no matter how we exit. Cool.
</blockquote></details>

Ack.


<!-- Sent from Reviewable.io -->

miretskiy commented:
@HonoreDB I changed backfill timestamp to be just a counter for SLI scope metrics.  That's because
the same scope could be used for multiple changefeeds.

stevendanna commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -142,6 +142,7 @@ type webhookMessage struct {
 type batch struct {
 	buffer      []messagePayload
 	bufferBytes int
+	messages    int
```
This looks unused other than in reset()

stevendanna commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -226,17 +207,88 @@ var (
+		Name: "commit_latency",
+		Help: "Event commit latency: a difference between event MVCC timestamp " +
+			"and the time it was acknowledged by the downstream sink.  If the sink batches events, " +
+			" then the difference between the oldest event in the back and acknowledgement is recorded; " +
```
s/back/batch/

miretskiy replied:
Done.

stevendanna reviewed (commented):
It is nice to see we are putting work into metrics.

We'll need to keep an eye on changes to the semantics of the mvcc_timestamp value (a topic of past discussion IIRC); however, I believe our current direction implies those will stay useful to us for these monitoring purposes.

I'm a tiny bit worried about exposing commit latency to end-users without more metrics about their sink's performance in order to fully understand the sources of the latency.  However, this also improves the flush metrics such that they are much more useful and provide a view into sink health. So, +1 on adding this now.

IIRC, in the case of batches, we are recording the longest commit latency in the batch since we don't want to have to store an arbitrary number of timestamps in the sink's state while we wait for the batch to send and then iterate over that arbitrary number of timestamps.  I think this is reasonable.  If we wanted to record more latency data in the future, perhaps we could move to some fixed-size buffer. Go's time.Time's are 24 bytes, so a 1MB fixed size buffer would allow us to keep ~41k timestamps.

miretskiy commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -22,79 +21,79 @@ import (
 		emitNanos := timeutil.Since(start).Nanoseconds()
-		s.metrics.EmittedMessages.Inc(1)
+		m.EmittedMessages.Inc(1)
 		// TODO(dan): This wasn't correct. The wrapped sink may emit the payload
 		// any number of times.
 		// s.metrics.EmittedBytes.Inc(int64(len(payload)))
```
I'm just going to drop this comment.  Don't think it's adding that much value.

miretskiy reviewed (commented):
I agree re keeping track of timestamps e.g. for webhook sink.  At this point, I'm willing to go with that -- it is still very useful for kafka sink for example.
We can improve this later (perhaps by keeping track of individual timestamps, as you suggest).

miretskiy commented:
@stevendanna @HonoreDB added additional check to return an error when sli_scope used w/out enterprise license.
Addressed all comments.  PTAL.

miretskiy commented:
@HonoreDB @stevendanna @amruss 
Please take another look at this PR.  U

HonoreDB reviewed (commented):
> Reviewed 4 of 16 files at r8, 10 of 20 files at r10, 13 of 13 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, @miretskiy, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_kafka.go, line 362 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoKQjWr6AD5CGYB9I2m:-MoKQjWr6AD5CGYB9I2n:b-s6nn7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/78da96b932927d0534d03a1c3f8d88bfc561694c/pkg/ccl/changefeedccl/sink_kafka.go#L362)):*
> ```Go
> 			if ackError == nil {
> 				m.updateMetrics(1, m.mvcc, ackMsg.Key.Length()+ackMsg.Value.Length(), sinkDoesNotCompress)
> 			}
> ```

Maybe safer to move the release after updateMetrics? That seems like it could theoretically get backed up and then we'd get untracked memory usage.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_kafka.go, line 362 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoKQjWr6AD5CGYB9I2m:-MoLAMn1BQ9Sxjy6keKC:b-4y9xil) ([raw file](https://github.com/cockroachdb/cockroach/blob/78da96b932927d0534d03a1c3f8d88bfc561694c/pkg/ccl/changefeedccl/sink_kafka.go#L362)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Maybe safer to move the release after updateMetrics? That seems like it could theoretically get backed up and then we'd get untracked memory usage.
</blockquote></details>

Don't thikn it matters, but can't hurt. Done
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 145 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MnGgUOCOe6sWNmI3Bvf-r6-145:-MoLA4Su4fiPt769L6LJ:b-uj46tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/53db00550dd9a99b4f2e3f2c04195eedcd125631/pkg/ccl/changefeedccl/sink_webhook.go#L145)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

This looks unused other than in reset()
</blockquote></details>

I think you probably meant some other bufffer bytes that I probably deleted at some point...
I think this buffer bytes is used to trigger flushes.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, @miretskiy, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLpJb40fkp7gCoKOgO:-MoLpJb40fkp7gCoKOgP:btq0opz) (no related file):*
Nice, removing the pre-defined scope names is a big improvement, as is the addition of a default scope and using the various Agg*  data types. It really cleans things up.  I left nits and notes as I was reading, but nothing blocking.

Reviewable has given up on me I think, so apologies in advance if my comments are connected to nonsensical lines.

---
___
*[-- commits, line 71 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLqATZBrPVJfcg-5-Z):*
I believe this is a cluster setting now.
___
*[-- commits, line 73 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLmfT107JlunsLCkYN):*
This still talks about tiers.  Perhaps "SLI scopes let operators label one or more changefeeds, based on criteria of their choosing."
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 529 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLo6rj4w_3tdKJVosp:-MoLo6rj4w_3tdKJVosq:bnq0ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/changefeed_processors.go#L529)):*
> ```Go
> 
> 	// Keep track of SLI latency for non-backfill/rangefeed KV events.
> 	if event.Type() == kvevent.TypeKV && event.BackfillTimestamp().IsEmpty() {
> ```

[nit] up to you, but if we move this into the case for TypeKV's below, then this function just has one place where we dispatch behavior based on type.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 342 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLs-S_6eBSj7B0_UV8:-MoLs-S_6eBSj7B0_UV9:b-y41lrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/changefeed_stmt.go#L342)):*
> ```Go
> 		telemetry.CountBucketed(`changefeed.create.num_tables`, int64(len(targets)))
> 
> 		if _, ok := opts[changefeedbase.OptSLIScope]; ok {
> ```

Should we validate somewhere that OptSLIScope that the user passed isn't the defaultSLIScope?  It wouldn't be a problem if it was, but it probably isn't what they intended?
___
*[pkg/ccl/changefeedccl/metrics.go, line 68 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLhgux-1_2Ss0-WC4g:-MoLhgux-1_2Ss0-WC4h:b9xg0vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L68)):*
> ```Go
> func (a *AggMetrics) MetricStruct() {}
> 
> // sliMetrics is a structure holding all SLI related metrics aggregated into AggMetrics.
> ```

[nit] `s/sliMetrics is a structure holding/sliMetrics holds/`, since we know it is a struct.
___
*[pkg/ccl/changefeedccl/metrics.go, line 115 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLjBZi7kSVSkupqjVO:-MoLjBZi7kSVSkupqjVP:br0ifpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L115)):*
> ```Go
> 	m.FlushedBytes.Inc(int64(compressedBytes))
> 	m.BatchHistNanos.RecordValue(emitNanos)
> 	if !m.backfilling.Get() {
> ```

Since the scanner pushes rows into a mem buffer, I believe that the scan can finish and the OnBackfillCallback func can be called before e all of those batches get emitted. In that case, this might record a few batches from the end of the backfill. Perhaps I've missed something that makes that impossible.

Even if the above issue is a problem, I'm not sure it is that big of a problem. Could be worth it to live with a few batches getting recorded to avoid having some more complex state to manage.
___
*[pkg/ccl/changefeedccl/metrics.go, line 264 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLkWrDCkqM2axu1TDj:-MoLkWrDCkqM2axu1TDk:b-iq5217) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L264)):*
> ```Go
> 		Help: "Event admission latency: a difference between event MVCC timestamp " +
> 			"and the time it was admitted into changefeed pipeline; " +
> 			"Note: this metric includes the time spent waiting until event   can be processed due " +
> ```

[nit] looks like some extra space here after "event"
___
*[pkg/ccl/changefeedccl/metrics.go, line 306 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLsX_D079Px0WXNApm:-MoLsX_D079Px0WXNApn:bo11kaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L306)):*
> ```Go
> }
> 
> func (a *AggMetrics) getOrCreateScope(scope string, sv *settings.Values) (*sliMetrics, error) {
> ```

I might be misunderstanding how the AggMetrics work, but to me it looks like the defaultSLIScope would cover all of the changefeeds _except_ those that have specified a scope.  If that is right, might be something we want to be careful to explain well when documenting this. 
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 145 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MnGgUOCOe6sWNmI3Bvf-r6-145:-MoLoNMfDkPO80iKWahy:bdm43fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/53db00550dd9a99b4f2e3f2c04195eedcd125631/pkg/ccl/changefeedccl/sink_webhook.go#L145)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think you probably meant some other bufffer bytes that I probably deleted at some point...
I think this buffer bytes is used to trigger flushes.
</blockquote></details>

Yeah, reviewable isn't making it very clear what this comment was referring to, but I don't see an issue now.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, and @stevendanna)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLpJb40fkp7gCoKOgO:-MoOYqXs9BUIhRmpKAM9:bb6sh2i) (no related file):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Nice, removing the pre-defined scope names is a big improvement, as is the addition of a default scope and using the various Agg*  data types. It really cleans things up.  I left nits and notes as I was reading, but nothing blocking.

Reviewable has given up on me I think, so apologies in advance if my comments are connected to nonsensical lines.


</blockquote></details>

Ack.

---
___
*[-- commits, line 71 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLqATZBrPVJfcg-5-Z):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I believe this is a cluster setting now.
</blockquote></details>

Ack.  Thanks.  Updated description.
___
*[-- commits, line 73 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLmfT107JlunsLCkYN):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

This still talks about tiers.  Perhaps "SLI scopes let operators label one or more changefeeds, based on criteria of their choosing."
</blockquote></details>

Ack.  Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 529 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLo6rj4w_3tdKJVosp:-MoO_RDNAnTOuF_SQiRv:b5srhs5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/changefeed_processors.go#L529)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] up to you, but if we move this into the case for TypeKV's below, then this function just has one place where we dispatch behavior based on type.
</blockquote></details>

Nice.  Definitely.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 342 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLs-S_6eBSj7B0_UV8:-MoOayXn9UxujwqEvsY_:bqayo5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/changefeed_stmt.go#L342)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Should we validate somewhere that OptSLIScope that the user passed isn't the defaultSLIScope?  It wouldn't be a problem if it was, but it probably isn't what they intended?
</blockquote></details>

Good idea.  Done.
___
*[pkg/ccl/changefeedccl/metrics.go, line 68 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLhgux-1_2Ss0-WC4g:-MoOb6QT8eNtq7oLJbx4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L68)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] `s/sliMetrics is a structure holding/sliMetrics holds/`, since we know it is a struct.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/metrics.go, line 115 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLjBZi7kSVSkupqjVO:-MoObJrq50Q0BGySyQfT:b28ihwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L115)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Since the scanner pushes rows into a mem buffer, I believe that the scan can finish and the OnBackfillCallback func can be called before e all of those batches get emitted. In that case, this might record a few batches from the end of the backfill. Perhaps I've missed something that makes that impossible.

Even if the above issue is a problem, I'm not sure it is that big of a problem. Could be worth it to live with a few batches getting recorded to avoid having some more complex state to manage.
</blockquote></details>

Right... Good catch...  I changed backfill timestamp impl.  PTAL.
___
*[pkg/ccl/changefeedccl/metrics.go, line 264 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLkWrDCkqM2axu1TDj:-MoOdpZ_El4xZ88wbkwq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L264)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] looks like some extra space here after "event"
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/metrics.go, line 306 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLsX_D079Px0WXNApm:-MoOdt-Y5N1dkgdwUZZy:bx3jfco) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L306)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I might be misunderstanding how the AggMetrics work, but to me it looks like the defaultSLIScope would cover all of the changefeeds _except_ those that have specified a scope.  If that is right, might be something we want to be careful to explain well when documenting this. 
</blockquote></details>

Sigh... That's right... Unfortunately.  I'll probably have to send another PR to update aggmetrics to enable the use case we want
(default scope aggregates all of them, while others don't)... Or... Perhaps we should make aggregators be the metrics themselves.
I'll ping this PR once that's done.
Good catch.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, and @stevendanna)

___
*[pkg/ccl/changefeedccl/metrics.go, line 306 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoLsX_D079Px0WXNApm:-MoOfGqO5ux8VJ57rZJN:b-53lhgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/fce1c899e6148905f913bc7423d0b01eb894dedc/pkg/ccl/changefeedccl/metrics.go#L306)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Sigh... That's right... Unfortunately.  I'll probably have to send another PR to update aggmetrics to enable the use case we want
(default scope aggregates all of them, while others don't)... Or... Perhaps we should make aggregators be the metrics themselves.
I'll ping this PR once that's done.
Good catch.
</blockquote></details>

Actually, that's a good catch... But I'm not sure I need to do anything.  All I need to do, I think is to rename my "aggregate" scope -- which I named before I got better understanding as to what this does, into "default" -- so that no implication are made wrt to
aggregation.  Aggregation happens at the top level agg metrics.  Those are the ones exported to external TSDB and they don't have any prometheus labels.  Everything named -- gets its own set of metrics; and everything else falls under default.
I think that's what we need.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, and @stevendanna)

___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 276 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoOjylQBLD4og4ICauS:-MoOjylQBLD4og4ICauT:bnwuiel) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd59e8628c03d222cf377fb139d51be73d14dfe5/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L276)):*
> ```Go
> 		// Regardless of whether we will actually execute backfill or not,
> 		// we will set backfill timestamp to the value of scanTime when this
> 		// function returns.
> ```

@stevendanna 
Just bringing this up so that I can have pair of eyes to double-check this logic.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, and @stevendanna)

___
*[pkg/ccl/changefeedccl/metrics.go, line 149 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoOkLyQ1tPRq-luiSTr:-MoOkLyQ1tPRq-luiSTs:bph66hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd59e8628c03d222cf377fb139d51be73d14dfe5/pkg/ccl/changefeedccl/metrics.go#L149)):*
> ```Go
> }
> 
> func (m *sliMetrics) getBackfillCallback() func() func(t hlc.Timestamp) {
> ```

@stevendanna 
FYI: moved this from changefeed_processors to keep backfill related logic closer to sliMetrics.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, and @stevendanna)

___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 276 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MoOjylQBLD4og4ICauS:-MoQ07HR9WM5yVf-xGzX:bt0a3jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd59e8628c03d222cf377fb139d51be73d14dfe5/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L276)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

@stevendanna 
Just bringing this up so that I can have pair of eyes to double-check this logic.
</blockquote></details>

Actually, I reverted this change... Basically, with the change to agg counters -- you're right, the only way to accurately
keep track of this is to pass in the event information down to the sink level (because events have backfill ts set).
Alas, this isn't the refactoring I care to do at this point -- we'd probably need to change kvevent.Alloc a bit or something similar.
At any rate, I simply check if backfill count is zero and keep track of latency only then.
We can revisit this later.


<!-- Sent from Reviewable.io -->

miretskiy commented:
@stevendanna @amruss  @HonoreDB 
I'm afraid more changes here.  Incorporated comments from our last meeting.
Namely:
  * Option renamed to `metrics_label`
  * Functionality controlled via a new env variable COCKROACH_EXPERIMENTAL_ENABLE_PER_CHANGEFEED_METRICS
  * I dropped support for a setting defining max number of scopes.  Instead, it's no hardcoded to 1024.
  
PTAL.  Want to get this merged into master and try to backport -- though I'm not confident.

HonoreDB reviewed (approved):
> Reviewed 15 of 15 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, @miretskiy, and @stevendanna)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 352 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MonjRz611GMDIptzufr:-MonjRz611GMDIptzufs:bvp13eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/863e153da8d872bd8a8a2b2adb52fdb3077258b5/pkg/ccl/changefeedccl/changefeed_stmt.go#L352)):*
> ```Go
> 			if scope == defaultSLIScope {
> 				return pgerror.Newf(pgcode.InvalidParameterValue,
> 					"%q=%q is the default metrics scope which keeps track of statistics "+
> ```

Nit (very nit): You could golf this using positional arguments: `%[1]q=%[2]q is the default`... and then `%[1]q` thereafter.
___
*[pkg/ccl/changefeedccl/metrics.go, line 38 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MonfKU3D0LJPBPLOxcN:-MonfKU3D0LJPBPLOxcO:bez46bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/863e153da8d872bd8a8a2b2adb52fdb3077258b5/pkg/ccl/changefeedccl/metrics.go#L38)):*
> ```Go
> 
> // defaultSLIScope is the name of the default SLI scope -- i.e. the set of metrics
> // keeping track of all cahngefeeds which did not have explicit sli scope specified.
> ```

typo: cahngefeeds
___
*[pkg/ccl/changefeedccl/metrics.go, line 337 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Monhg3I6qu-gTI2GNve:-Monhg3I6qu-gTI2GNvf:b-25v1mx) ([raw file](https://github.com/cockroachdb/cockroach/blob/863e153da8d872bd8a8a2b2adb52fdb3077258b5/pkg/ccl/changefeedccl/metrics.go#L337)):*
> ```Go
> 
> 	const failSafeMax = 1024
> 	if len(a.mu.sliMetrics) > failSafeMax {
> ```

This could let me get into kind of a weird state if I create 1025 labels before trying to use the default; I now can't use the default and the number in the error message is off by one. Maybe using the default scope should just skip all these validations, and/or you could just ensure the default scope is always created (maybe you're already doing that but if so you have dead code above.)


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @HonoreDB, and @stevendanna)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 352 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MonjRz611GMDIptzufr:-MooQRTa0Ki3FK1mN_id:bw1m75j) ([raw file](https://github.com/cockroachdb/cockroach/blob/863e153da8d872bd8a8a2b2adb52fdb3077258b5/pkg/ccl/changefeedccl/changefeed_stmt.go#L352)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Nit (very nit): You could golf this using positional arguments: `%[1]q=%[2]q is the default`... and then `%[1]q` thereafter.
</blockquote></details>

Definitely! Thanks!
___
*[pkg/ccl/changefeedccl/metrics.go, line 38 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-MonfKU3D0LJPBPLOxcN:-MooRHBM298saSnEw0EY:b2f6e7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/863e153da8d872bd8a8a2b2adb52fdb3077258b5/pkg/ccl/changefeedccl/metrics.go#L38)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

typo: cahngefeeds
</blockquote></details>

Tahkn yuo
___
*[pkg/ccl/changefeedccl/metrics.go, line 337 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/72111#-Monhg3I6qu-gTI2GNve:-MooRzrv3OUs3a3DPIkp:b4knqr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/863e153da8d872bd8a8a2b2adb52fdb3077258b5/pkg/ccl/changefeedccl/metrics.go#L337)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

This could let me get into kind of a weird state if I create 1025 labels before trying to use the default; I now can't use the default and the number in the error message is off by one. Maybe using the default scope should just skip all these validations, and/or you could just ensure the default scope is always created (maybe you're already doing that but if so you have dead code above.)
</blockquote></details>

Defense is good.  Done.


<!-- Sent from Reviewable.io -->

miretskiy commented:
bors r+

miretskiy commented:
bors r+
